Enable GPU acceleration and parallelization across Dynearthsol

chaseshyu opened the pull request:
Enable GPU acceleration and parallelization across Dynearthsol

Features:
- Enable Nvidia GPU support with async/await CUDA execution
- Add new 3D initial condition: dome-shaped temperature anomaly (`temperature_option = 3`)
- Support compressed NetCDF4 output format (compile with `netcdf=1`)
- Introduce new marker array: marker_in_elem for parallelized marker operations
- Add param.mat.mattype_ref for calculate_dt(), ref_pressure(), and compute_mass()
- Add var.contact and var.neighbor to store element facet contact neighbors
- Add parallelized 2vtk.py conversion option (`-ncpu #`)

Performance:
- Newly parallelize routines:
  - remesh(), ANN search, apply_stress_bcs(), create_boundary_facets()
  - phase_changes() and surface_processes()
- Use nanoflann for CPU KNN search
- Use spatial hash grid CUDA kernel for GPU KNN search
- Use precomputed tables for log(), sin(), tan() when `nprof=1`
- Force mat array size to param.mat.nmat for better material property calculations

Build/Infra:
- Append .gpu suffix to object and execution files when make `openacc=1`
- GPU A100 + 4 CPU cores benchmark: 1M elements / 5 Myr in 13 hrs (3x faster than CPU 32 cores; see `benchmarks-cores/test-3d-equ-long`)

Bug Fixes:
- Fix memory leak in array2D::resize()
- Fix using wrong old coord for remeshing plane interpolation

chaseshyu commented on markerset.cxx:
`edvacc` means the accumulated volume changes of elements due to certain process on the surface. In present version, the process only is surface sediment deposition caused by topographic diffusion.

chaseshyu commented on markerset.cxx:
The program actually will stop (`std::exit(1)`) if any condition matches and makes `++is_error`.

chaseshyu replied:
Yes, you can compare the changes with the master branch to see the overall changes of this merge. Sometimes, I find logical errors in a modification after a few tests and commits. So, I make a bugfixes commit later.

chaseshyu commented on bc.cxx:
The parallelized loop involving atomic operations on float numbers could have varying orders of number operations, potentially leading to unexpected round-off errors that undermine DES’s deterministics. While a two-stage parallelized loop method might address this issue, the complexity of this loop here surpasses that of a simple element-node loop. More time is needed to implement a two-stage method here.

chaseshyu commented on nn-interpolation.cxx:
Performance. find_acm_elem_ratios() now is pararllized.
1. Reference points for element ratio are created before the loop of find nn.
2. Avoid tricky reference point on the element edges (eta = 0).
3. Avoid vector push back on GPU.

chaseshyu commented on barycentric-fn.hpp:
It's a good question. I want to ask NVIDIA also ...XD
NV compiler would not parallelize the loop if the type of loop index is not INT.
So I need to use INT here for loop in this class.

chaseshyu commented on rheology.cxx:
Here is to avoid allocation of elem#- or node#-length temporary arrays (size can be several millions of doubles) in every time step.

chaseshyu commented on fields.cxx:
Here is a temporary disable for make sure deterministics of DES. I have recovered acc later.

echoi commented on fields.cxx:
Sorry, I raised this issue but don't have a firm opinion about whether to store `tdot`. Would it be an option to keep only the `dt * tdot / tmass` structure exposed, not saving `tdot` in another array?
Or, how about keeping your change while adding the original code as a comment?

chaseshyu replied:
Sure, I will rename `tmp` to `tdot`.

https://github.com/GeoFLAC/DynEarthSol/blob/b166734f28d8d5dc0da7c832bfd1582db55a55be/fields.cxx#L213

chaseshyu commented on utils.hpp:
```diff
@@ -67,6 +64,106 @@ void print(std::ostream& os, const Array& A)
+#pragma acc routine seq
+static inline double log_lookup(const double_vec& log_table, double x) {
+    if (x <= 0.0)
+        printf("Error: log_lookup called with x <= 0 (%g)\n", x);
```
https://github.com/GeoFLAC/DynEarthSol/blob/aee7bffadc781fafb94d7fab4df30b9c0b0065a0/utils.hpp#L71-L73

https://github.com/GeoFLAC/DynEarthSol/blob/aee7bffadc781fafb94d7fab4df30b9c0b0065a0/utils.hpp#L93-L95

https://github.com/GeoFLAC/DynEarthSol/blob/aee7bffadc781fafb94d7fab4df30b9c0b0065a0/utils.hpp#L106-L108

echoi commented:
@chaseshyu Thanks for addressing the review comments, even those from Copilot! I have marked most of my comments as resolved. Please do the same for the remaining ones if you think they can be closed. When all are resolved, I'll approve this PR.

chaseshyu commented:
> @chaseshyu Thanks for addressing the review comments, even those from Copilot! I have marked most of my comments as resolved. Please do the same for the remaining ones if you think they can be closed. When all are resolved, I'll approve this PR.

@echoi No problem! Thank you for putting so much effort into reviewing the code. I have all resolved comments.

tan2 commented:
Sorry to join the discussion late. I am back from the field.

In the early version, when nmat=3 but an input file has only rho0=[2800],
p.mat.rho0 is resized to 3 and all three slots are filled with 2800, like
Eunseo described.

Sometime in 2013(?), I profiled the code with  nmat=20, and found that
indexing of all material properties would take a significant amount of time
(~10% of time, from my memory). As an optimization, I wrote a new class (I
forgot its name), which has the same interface as a vector but, when
indexed, always returns a constant value. Some material properties, such as
heat capacity and thermal conductivity, are commonly constant for all
materials. This class can significantly reduce the indexing time and speed
up the code.

The new code seems to have reverted this optimization. Since we rarely use
a large number of materials in the models, my optimization might be
unnecessary and over-complicate the code.


On Mon, Aug 25, 2025 at 9:56 PM Chase J. Shyu ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On input.cxx
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#discussion_r2298183059>:
>
> Right. It was [2800], Now it is [2800, 2800, 2800].
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#discussion_r2298183059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABWQQT4MSPWIIRWERBGON33PMIXTAVCNFSM6AAAAACEFYBMXWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCNJRGU4TEMBTGI>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

chaseshyu commented:
@echoi @tan2 Thanks for following up. If it reduces the efficiency, reverting it back to the original is a good idea. I can just write another function to deal with this inconsistency among different mat sizes.

echoi commented:
Let's use this opportunity to discuss how we handle implicit intentions
regarding `mat` array sizes.

Here's my thinking on different cases:
* **Case 1: `mat` array size == `nmat`**: This scenario has no ambiguity
and should be the standard.
* **Case 2: `mat` array size = 1, `nmat` > 1**: The user's intention to
populate the `mat` array with a single value is clear and should be
permitted.
* **Case 3: `mat` array size > 1 and != `nmat`**: This situation is
ambiguous and could indicate a mistake. I believe an error should be thrown
in this case to prevent unintended behavior.

Are there any other legitimate combinations we should consider?


On Sun, Aug 31, 2025 at 2:15 PM Chase J. Shyu ***@***.***>
wrote:

> *chaseshyu* left a comment (GeoFLAC/DynEarthSol#25)
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240354285>
>
> @echoi <https://github.com/echoi> @tan2 <https://github.com/tan2> Thanks
> for following up. If it reduces the efficiency, reverting it back to the
> original is a good idea. I can just write another function to deal with
> this inconsistency among different mat sizes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240354285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJKUNW3CWMOMLBSDPVUIRL3QNCWNAVCNFSM6AAAAACEFYBMXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENBQGM2TIMRYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tan2 commented:
Chase,

It seems that the VectorBase class is removed while adding openacc support.
Is there any difficulty to convert the class with openacc?


On Mon, Sep 1, 2025 at 3:15 AM Chase J. Shyu ***@***.***>
wrote:

> *chaseshyu* left a comment (GeoFLAC/DynEarthSol#25)
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240354285>
>
> @echoi <https://github.com/echoi> @tan2 <https://github.com/tan2> Thanks
> for following up. If it reduces the efficiency, reverting it back to the
> original is a good idea. I can just write another function to deal with
> this inconsistency among different mat sizes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240354285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABWQQSS7CZ2RSABDLU5NDL3QNCWNAVCNFSM6AAAAACEFYBMXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENBQGM2TIMRYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tan2 commented:
I agree that Case 3 should be an error. I will implement it.

On Mon, Sep 1, 2025 at 3:35 AM Eunseo Choi ***@***.***> wrote:

> *echoi* left a comment (GeoFLAC/DynEarthSol#25)
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240363920>
> Let's use this opportunity to discuss how we handle implicit intentions
> regarding `mat` array sizes.
>
> Here's my thinking on different cases:
> * **Case 1: `mat` array size == `nmat`**: This scenario has no ambiguity
> and should be the standard.
> * **Case 2: `mat` array size = 1, `nmat` > 1**: The user's intention to
> populate the `mat` array with a single value is clear and should be
> permitted.
> * **Case 3: `mat` array size > 1 and != `nmat`**: This situation is
> ambiguous and could indicate a mistake. I believe an error should be
> thrown
> in this case to prevent unintended behavior.
>
> Are there any other legitimate combinations we should consider?
>
>
> On Sun, Aug 31, 2025 at 2:15 PM Chase J. Shyu ***@***.***>
> wrote:
>
> > *chaseshyu* left a comment (GeoFLAC/DynEarthSol#25)
> > <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240354285>
>
> >
> > @echoi <https://github.com/echoi> @tan2 <https://github.com/tan2>
> Thanks
> > for following up. If it reduces the efficiency, reverting it back to the
> > original is a good idea. I can just write another function to deal with
> > this inconsistency among different mat sizes.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240354285>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABJKUNW3CWMOMLBSDPVUIRL3QNCWNAVCNFSM6AAAAACEFYBMXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENBQGM2TIMRYGU>
>
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GeoFLAC/DynEarthSol/pull/25#issuecomment-3240363920>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABWQQTVLTN3KVSDDLMDZ3T3QNFBXAVCNFSM6AAAAACEFYBMXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENBQGM3DGOJSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
